compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalCompilerRunner.kt

```diff
@@ -193,37 +193,42 @@ abstract class IncrementalCompilerRunner<
             rebuild(BuildAttribute.CACHE_CORRUPTION)
         } finally {
             if (cachesMayBeCorrupted) {
-                clearLocalStateOnRebuild(args)
+                clearOutputsOnRebuild(args)
             }
         }
     }
 
-    private fun clearLocalStateOnRebuild(args: Args) {
+    /**
+     * Deletes output files and contents of output directories on rebuild (including `@LocalState` files/directories).
+     *
+     * If the output directories do not yet exist, they will be created.
+     */
+    private fun clearOutputsOnRebuild(args: Args) {
         val destinationDir = destinationDir(args)
-
-        reporter.reportVerbose { "Clearing output on rebuild" }
-        for (file in sequenceOf(destinationDir, workingDir) + outputFiles.asSequence()) {
-            val deleted: Boolean? = when {
-                file.isDirectory -> {
-                    reporter.reportVerbose { "  Deleting directory $file" }
-                    file.deleteRecursively()
-                }
-                file.isFile -> {
-                    reporter.reportVerbose { "  Deleting $file" }
-                    file.delete()
-                }
-                else -> null
+        val (outputDirsThatExist, regularOrNonExistentOutputFiles) = outputFiles.partition { it.isDirectory }
+        val regularOutputFiles = regularOrNonExistentOutputFiles.filter { it.exists() }
+
+        // outputDirsThatExist may or may not contain destinationDir and workingDir.
+        // Collect all of them so that we don't miss any output directories.
+        // Use Set to avoid duplication.
+        val allOutputDirs = setOf(destinationDir, workingDir) + outputDirsThatExist
+
+        reporter.reportVerbose { "Clearing outputs on rebuild" }
+        allOutputDirs.forEach { dir ->
+            reporter.reportVerbose { "  Deleting contents of directory '${dir.path}'" }
+            dir.listFiles()?.forEach {
+                it.deleteRecursively()
+                if (it.exists()) throw IOException("Could not delete '${it.path}'")
             }
 
-            if (deleted == false) {
-                reporter.reportVerbose { "  Could not delete $file" }
-            }
+            dir.mkdirs()
```

Comment:
If the directory already exists, `mkdirs` will return `false`, so it will be unexpected. Anyway, in the new version, I use your suggestion but ensure the directory does not exist first.

Reply:
LGTM